### PR TITLE
Flowtest: Test base operation "node drain", followed by other operations.

### PR DESCRIPTION
**Pull request by parikshithb**
Test covers following flow operations while running workloads in the background:
        1. Node drain
        2. Add capacity
        3. Node reboot
        4. Node n/w failure

Signed-off-by: Parikshith <pbyregow@redhat.com>

### CODE REVIEW

File: tests/e2e/flowtest/test_base_operation_node_drain.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+@ignore_leftovers
+@tier2
+class TestFlowBasedTests(E2ETest):
+    """
+    Tests Story/Flow based test scenarios
+
+    """
+    @vsphere_platform_required
+    @aws_platform_required
+    @pytest.mark.polarion_id("OCS-xsadsa")
```
Comment (hnallurv) [inline tests/e2e/flowtest/test_base_operation_node_drain.py:28]: Add actual polarion id

File: tests/e2e/flowtest/test_base_operation_node_drain.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        logger.info("Starting IO operations in Background")
+        project = project_factory()
+
+        bg_handler = flowtest_helpers.BackgroundOps()
+        executor_run_bg_ios_ops = ThreadPoolExecutor(max_workers=3)
+
+        logging.info("Started object IOs in background")
+        obc_ios = executor_run_bg_ios_ops.submit(
+            bg_handler.handler, obc_put_obj_create_delete, mcg_obj, bucket_factory, iterations=50
+        )
```
Comment (hnallurv) [inline tests/e2e/flowtest/test_base_operation_node_drain.py:50]: move logging.info("Started object IOs in background") to after this line.

File: tests/e2e/flowtest/test_base_operation_node_drain.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        logging.info("Started pvc create and delete in background")
+        pvc_create_delete = executor_run_bg_ios_ops.submit(
+            bg_handler.handler,
+            create_pvc_delete,
+            multi_pvc_factory,
+            project,
+            iterations=100
+        )
```
Comment (hnallurv) [inline tests/e2e/flowtest/test_base_operation_node_drain.py:58]: move logging.info("Started pvc create and delete in background") to after this line

File: tests/e2e/flowtest/test_base_operation_node_drain.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        logging.info("Started pgsql workload in background")
+        pgsql_workload = executor_run_bg_ios_ops.submit(
+            bg_handler.handler,
+            pgsql_factory_fixture,
+            replicas=1, clients=1, transactions=100,
+            timeout=100, iterations=1
+        )
+
```
Comment (hnallurv) [inline tests/e2e/flowtest/test_base_operation_node_drain.py:66]: move  logging.info("Started pgsql workload in background") to this line

File: tests/e2e/flowtest/flowtest_helpers.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+        Initialize Sanity instance
+
+        """
+        self.sanity_helpers = Sanity()
+
+    def node_operations_entry_criteria(self, node_type, number_of_nodes, network_fail_time=None):
+        """
+        Entry criteria function for node related operations
+        Args:
+            node_type (str): Tyoe of node
```
Comment (ebenahar) [inline tests/e2e/flowtest/flowtest_helpers.py:31]: `Type`

File: tests/e2e/flowtest/flowtest_helpers.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+            number_of_nodes (int): Number of nodes
+            network_fail_time (int): Total time to fail the network in a node
```
Comment (ebenahar) [inline tests/e2e/flowtest/flowtest_helpers.py:33]: please add an empty line below

File: tests/e2e/flowtest/flowtest_helpers.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+    def __init__(self):
+        """
+        Initialize Sanity instance
+
+        """
+        self.sanity_helpers = Sanity()
+
+    def node_operations_entry_criteria(self, node_type, number_of_nodes, network_fail_time=None):
+        """
+        Entry criteria function for node related operations
```
Comment (ebenahar) [inline tests/e2e/flowtest/flowtest_helpers.py:29]: please add an empty line below

File: tests/e2e/flowtest/flowtest_helpers.py [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+
+class FlowOperations:
+    def __init__(self):
+        """
+        Initialize Sanity instance
+
+        """
+        self.sanity_helpers = Sanity()
+
+    def node_operations_entry_criteria(self, node_type, number_of_nodes, network_fail_time=None):
```
Comment (ebenahar) [inline tests/e2e/flowtest/flowtest_helpers.py:27]: these functions seem relevant for other tests as well

File: tests/e2e/flowtest/flowtest_helpers.py [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+        logger.info("Node operations exit: Verifying whether node is ready")
+        wait_for_nodes_status(status=constants.NODE_READY, timeout=300)
+        logger.info("Node operations exit: Verifying cluster health")
+        self.sanity_helpers.health_check(cluster_check=False, tries=80)
+        logger.info("Node operations exit: Verifying StorageCluster pods are in running/completed state")
+        wait_for_storage_pods(timeout=300), 'Some pods were not in expected state'
+
+        logging.info("Node operation: Exit criteria verification: Success")
+
+    def add_capacity_entry_criteria(self):
```
Comment (ebenahar) [inline tests/e2e/flowtest/flowtest_helpers.py:71]: I think we have something similar under tests.manage.z_cluster.cluster_expansion.test_add_capacity_entry_exit_criteria.TestAddCapacity.test_add_capacity

File: tests/e2e/flowtest/flowtest_helpers.py [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+        number_of_osds_added = len(osd_pods_after) - len(osd_pods_before)
+        logger.info(f"Number of osds added = {number_of_osds_added}, "
+                    f"before = {len(osd_pods_before)}, after = {len(osd_pods_after)}")
+        assert number_of_osds_added == 3, "Exit criteria verification FAILED: osd count mismatch"
+
+        logging.info("Add capacity: Exit criteria verification: Success")
+
+
+@ignore_leftovers
+def create_pvc_delete(multi_pvc_factory, project=None):
```
Comment (ebenahar) [inline tests/e2e/flowtest/flowtest_helpers.py:123]: I think this function should be in sanity_helpers

File: tests/e2e/flowtest/flowtest_helpers.py [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+
+    # Check PVCs are deleted
+    for pvc_obj in all_pvc_to_delete:
+        pvc_obj.ocp.wait_for_delete(resource_name=pvc_obj.name)
+
+    logger.info("All PVCs are deleted as expected")
+    return True
+
+
+def obc_put_obj_create_delete(mcg_obj, bucket_factory):
```
Comment (ebenahar) [inline tests/e2e/flowtest/flowtest_helpers.py:161]: I think this function should be in sanity_helpers

File: tests/e2e/flowtest/flowtest_helpers.py [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+        Entry criteria function for node related operations
+        Args:
+            node_type (str): Tyoe of node
+            number_of_nodes (int): Number of nodes
+            network_fail_time (int): Total time to fail the network in a node
+        Returns:
+            tuple: containing the params used in Node operations
+
+        """
+        logger.info("Node operations entry: Verifying cluster health")
```
Comment (hnallurv) [inline tests/e2e/flowtest/flowtest_helpers.py:38]: The lines 38 to 43 are common to both node_operations_entry_criteria() and node_operations_exit_criteria. Can they be under one generic function?..

File: tests/e2e/flowtest/flowtest_helpers.py [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+    data = "A string data"
+
+    for i in range(0, 30):
+        key = 'Object-key-' + f"{i}"
+        logger.info(f"Write, read and delete object with key: {key}")
+        assert s3_put_object(mcg_obj, bucket_name, key, data), f"Failed: Put object, {key}"
+        assert s3_get_object(mcg_obj, bucket_name, key), f"Failed: Get object, {key}"
+        assert s3_delete_object(mcg_obj, bucket_name, key), f"Failed: Delete object, {key}"
+
+    return True
```
Comment (hnallurv) [inline tests/e2e/flowtest/flowtest_helpers.py:178]: this may not be needed. please check

File: tests/e2e/flowtest/test_base_operation_node_drain.py (hunk 1 above)
Comment (parikshithb) [reply tests/e2e/flowtest/test_base_operation_node_drain.py]: Ack

File: tests/e2e/flowtest/test_base_operation_node_drain.py (hunk 2 above)
Comment (parikshithb) [reply tests/e2e/flowtest/test_base_operation_node_drain.py:48]: ack

File: tests/e2e/flowtest/test_base_operation_node_drain.py (hunk 3 above)
Comment (parikshithb) [reply tests/e2e/flowtest/test_base_operation_node_drain.py:57]: ack

File: tests/e2e/flowtest/test_base_operation_node_drain.py (hunk 4 above)
Comment (parikshithb) [reply tests/e2e/flowtest/test_base_operation_node_drain.py:67]: ack

File: tests/e2e/flowtest/flowtest_helpers.py (hunk 5 above)
Comment (parikshithb) [reply tests/e2e/flowtest/flowtest_helpers.py]: ack

File: tests/e2e/flowtest/flowtest_helpers.py (hunk 6 above)
Comment (parikshithb) [reply tests/e2e/flowtest/flowtest_helpers.py:60]: ack

File: tests/e2e/flowtest/flowtest_helpers.py (hunk 7 above)
Comment (parikshithb) [reply tests/e2e/flowtest/flowtest_helpers.py:54]: ack

File: tests/e2e/flowtest/flowtest_helpers.py (hunk 13 above)
Comment (parikshithb) [reply tests/e2e/flowtest/flowtest_helpers.py]: ack. Removed

File: tests/e2e/flowtest/flowtest_helpers.py (hunk 12 above)
Comment (parikshithb) [reply tests/e2e/flowtest/flowtest_helpers.py]: Created a generic function called validate cluster

File: tests/e2e/flowtest/flowtest_helpers.py (hunk 11 above)
Comment (parikshithb) [reply tests/e2e/flowtest/flowtest_helpers.py:166]: There is one in mcg helpers but it uses awscli pod to write all data, if node running awscli pod is interrupted IOs will stop. This function will just use boto3.

File: tests/e2e/flowtest/flowtest_helpers.py (hunk 10 above)
Comment (parikshithb) [reply tests/e2e/flowtest/flowtest_helpers.py:128]: I couldn't find a proper one, there are functions similar to this but it also creates pods as well. My use case is just pvc create and delete. Also Pods will be interrupted during node failure operations.

File: tests/e2e/flowtest/flowtest_helpers.py (hunk 9 above)
Comment (parikshithb) [reply tests/e2e/flowtest/flowtest_helpers.py:80]: True, but they are part of the test case itself and are scattered, this is a generic function.

File: tests/conftest.py [hunk 14]
```diff
@@ -2683,3 +2684,67 @@ def ceph_toolbox(request):
...
+        if scheduling_disabled_nodes:
+            schedule_nodes(scheduling_disabled_nodes)
+
+        # Remove label created for DC app pods on all worker nodes
+        node_objs = get_node_objs()
+        for node_obj in node_objs:
+            if 'dc' in node_obj.get().get('metadata').get('labels').keys():
+                remove_label_from_worker_node(
+                    [node_obj.name], label_key="dc"
+                )
```
Comment (PrasadDesala) [inline tests/conftest.py:2715]: These lines doesn't seem related to node drain tear_down. 
We are removing the label created for DC app pod here.
Any reason for adding this?

File: tests/conftest.py [hunk 15]
```diff
@@ -2683,3 +2684,67 @@ def ceph_toolbox(request):
...
+            if not_ready_nodes:
+                log.info(
+                    f"Nodes in NotReady status found: {[n.name for n in not_ready_nodes]}"
+                )
+                nodes.restart_nodes(not_ready_nodes)
+                node.wait_for_nodes_status(status=constants.NODE_READY)
+
+        # Check ceph health
+        assert ceph_health_check(tries=80), "Ceph cluster health is not OK"
+        log.info("Ceph cluster health is OK")
```
Comment (PrasadDesala) [inline tests/conftest.py:2748]: Can the ceph health be handled separately? this seems out of scope to the node_restart_teardown().
WDYT?

File: tests/e2e/flowtest/flowtest_helpers.py [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+
+from ocs_ci.ocs import node, platform_nodes, defaults, exceptions
+from ocs_ci.ocs.node import wait_for_nodes_status
+from ocs_ci.ocs.resources import pod as pod_helpers
+from ocs_ci.ocs.resources.pod import wait_for_storage_pods
+from ocs_ci.utility.utils import TimeoutSampler
+from tests.manage.mcg.helpers import s3_delete_object, s3_put_object, s3_get_object
+from tests.sanity_helpers import Sanity
+from ocs_ci.framework.testlib import ignore_leftovers
+from ocs_ci.ocs import constants
```
Comment (PrasadDesala) [inline tests/e2e/flowtest/flowtest_helpers.py:11]: Can be imported in line#3

File: tests/e2e/flowtest/flowtest_helpers.py [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+
+        """
+        self.validate_cluster(node_status=True, operation_name=operation_name)
+
+        logger.info(f"Getting parameters related to: {operation_name}")
+        typed_nodes = node.get_typed_nodes(
+            node_type=node_type, num_of_nodes=number_of_nodes
+        )
+        factory = platform_nodes.PlatformNodesFactory()
+        nodes = factory.get_nodes_platform()
```
Comment (PrasadDesala) [inline tests/e2e/flowtest/flowtest_helpers.py:73]: nodes instance is already created in confest.py, do we need to create it again?

File: tests/e2e/flowtest/test_base_operation_node_drain.py [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+from concurrent.futures import ThreadPoolExecutor
+
+from ocs_ci.framework.pytest_customization.marks import aws_platform_required, vsphere_platform_required
+from ocs_ci.ocs import node, constants
+from ocs_ci.framework.testlib import E2ETest, tier2, config, ignore_leftovers
+from ocs_ci.ocs.ocp import OCP
+from ocs_ci.ocs.resources import storage_cluster
+from tests.e2e.flowtest import flowtest_helpers
+from tests.e2e.flowtest.flowtest_helpers import create_pvc_delete, \
```
Comment (PrasadDesala) [inline tests/e2e/flowtest/test_base_operation_node_drain.py:13]: Please break the line with ()

File: tests/e2e/flowtest/test_base_operation_node_drain.py [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+        logger.info("Starting operation 1: Node Drain")
+        node_name, nodes = flow_ops.node_operations_entry_criteria(
+            node_type='worker', number_of_nodes=1, operation_name="Node Drain"
+        )
+        # Node maintenance - to gracefully terminate all pods on the node
+        node.drain_nodes([node_name[0].name])
+        # Make the node schedulable again
+        node.schedule_nodes([node_name[0].name])
+        logger.info("Verifying exit criteria for operation 1: Node Drain")
+        flow_ops.validate_cluster(node_status=True, operation_name="Node Drain")
```
Comment (PrasadDesala) [inline tests/e2e/flowtest/test_base_operation_node_drain.py:78]: After scheduling back the node, we should verify that all the pod are in running state.
Please add arg pod_status to true

File: tests/e2e/flowtest/test_base_operation_node_drain.py [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+            kind=constants.POD, namespace=config.ENV_DATA['cluster_namespace']
+        )
+        pod.wait_for_resource(
+            timeout=300,
+            condition=constants.STATUS_RUNNING,
+            selector='app=rook-ceph-osd',
+            resource_count=result * 3
+        )
+        logger.info("Verifying exit criteria for operation 2: Add Capacity")
+        flow_ops.add_capacity_exit_criteria(restart_count_before, osd_pods_before)
```
Comment (PrasadDesala) [inline tests/e2e/flowtest/test_base_operation_node_drain.py:95]: We should also validate the cluster state after adding a capacity.
May be we can add this check in add_capcity exit criteria?

Comment same for other ops also, basically we should check the system is healthy for every operation.

File: tests/e2e/flowtest/test_base_operation_node_drain.py [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+        flow_ops.validate_cluster(node_status=True, operation_name="Node Restart")
+
+        logger.info("Starting operation 4: Node network fail")
+        node_name, node_platform, nw_fail_time = flow_ops.node_operations_entry_criteria(
+            node_type='worker', number_of_nodes=1, network_fail_time=300, operation_name="Node N/W failure"
+        )
+        # Node n/w interface failure
+        node.node_network_failure(node_name[0].name)
+        logger.info(f"Waiting for {nw_fail_time} seconds")
+        sleep(nw_fail_time)
```
Comment (PrasadDesala) [inline tests/e2e/flowtest/test_base_operation_node_drain.py:113]: Do we need these?
In the node_network_failure(), when the network is failed on the node, we are waiting for it to reach not ready state.

File: tests/e2e/flowtest/flowtest_helpers.py (hunk 10 above)
Comment (ebenahar) [reply tests/e2e/flowtest/flowtest_helpers.py:128]: I meant that we should move this function to be a sanity_helpers method

File: tests/e2e/flowtest/flowtest_helpers.py (hunk 11 above)
Comment (ebenahar) [reply tests/e2e/flowtest/flowtest_helpers.py:166]: I meant that we should move this function to be a sanity_helpers method

File: tests/e2e/flowtest/flowtest_helpers.py (hunk 10 above)
Comment (parikshithb) [reply tests/e2e/flowtest/flowtest_helpers.py]: Thanks, moved

File: tests/e2e/flowtest/flowtest_helpers.py (hunk 11 above)
Comment (parikshithb) [reply tests/e2e/flowtest/flowtest_helpers.py]: moved

File: tests/conftest.py (hunk 14 above)
Comment (parikshithb) [reply tests/conftest.py]: Nope no reason. Removed

File: tests/conftest.py (hunk 15 above)
Comment (parikshithb) [reply tests/conftest.py]: In my testing due to amount of data there will me be minimal rebalancing happening after node restart, i think it is better we validate health as well in node teardown.

File: tests/e2e/flowtest/flowtest_helpers.py (hunk 16 above)
Comment (parikshithb) [reply tests/e2e/flowtest/flowtest_helpers.py]: ack

File: tests/e2e/flowtest/flowtest_helpers.py (hunk 17 above)
Comment (parikshithb) [reply tests/e2e/flowtest/flowtest_helpers.py]: ack

File: tests/e2e/flowtest/test_base_operation_node_drain.py (hunk 18 above)
Comment (parikshithb) [reply tests/e2e/flowtest/test_base_operation_node_drain.py]: ack

File: tests/e2e/flowtest/test_base_operation_node_drain.py (hunk 19 above)
Comment (parikshithb) [reply tests/e2e/flowtest/test_base_operation_node_drain.py]: Added the check

File: tests/e2e/flowtest/test_base_operation_node_drain.py (hunk 20 above)
Comment (parikshithb) [reply tests/e2e/flowtest/test_base_operation_node_drain.py:99]: It is validated in add capacity exit as well, checks are validated inside the add capacity exit criteria method.

File: tests/e2e/flowtest/test_base_operation_node_drain.py (hunk 21 above)
Comment (parikshithb) [reply tests/e2e/flowtest/test_base_operation_node_drain.py:117]: It is part of the test case which has two scenarios: 1. short down time after n/w failure 2. prolonged down time after n/w failure

Comment (parikshithb) [reply tests/e2e/flowtest/test_base_operation_node_drain.py:117]: So decided to go with short wait time.

File: tests/e2e/flowtest/flowtest_helpers.py [hunk 22]
```diff
@@ -0,0 +1,165 @@
...
+        osd_pods_after = pod_helpers.get_osd_pods()
+        number_of_osds_added = len(osd_pods_after) - len(osd_pods_before)
+        logger.info(f"Number of OSDs added = {number_of_osds_added}, "
+                    f"before = {len(osd_pods_before)}, after = {len(osd_pods_after)}")
+        assert number_of_osds_added == 3, "Exit criteria verification FAILED: osd count mismatch"
+
+        logger.info("Add capacity: Exit criteria verification: Success")
+
+
+class BackgroundOps:
```
Comment (ebenahar) [inline tests/e2e/flowtest/flowtest_helpers.py:118]: very nice!!

File: tests/e2e/flowtest/flowtest_helpers.py [hunk 23]
```diff
@@ -0,0 +1,165 @@
+import logging
```
Comment (ebenahar) [inline tests/e2e/flowtest/flowtest_helpers.py:1]: we should try to avoid having helper files. Will it make sense to move this to be under ocs_ci/?

Comment (parikshithb) [reply tests/e2e/flowtest/flowtest_helpers.py:1]: There are currently many test cases planned for flow/story based scenarios, so i think a dedicated helper file might be useful. @PrasadDesala @hnallurv What is your opinion on this ?

Comment (ebenahar) [reply tests/e2e/flowtest/flowtest_helpers.py:1]: that's understood. What I meant is that we should move the entire file to be placed under ocs_ci/ hierarchy.

Comment (PrasadDesala) [reply tests/e2e/flowtest/flowtest_helpers.py:1]: +1 to move "flowtest_helpers.py " under ocs_ci/

Comment (parikshithb) [reply tests/e2e/flowtest/flowtest_helpers.py]: ack. Moved
